File: linux/src/stream_options.cpp
#include "stream_options.hpp"

#include <cstdint>
#include <vector>

namespace
{
    void writeInt32(
        std::vector<std::uint8_t>& buffer,
        std::int32_t value)
    {
        const std::uint32_t unsignedValue =
            static_cast<std::uint32_t>(value);

        buffer.push_back(
            static_cast<std::uint8_t>((unsignedValue >> 24) & 0xFF));

        buffer.push_back(
            static_cast<std::uint8_t>((unsignedValue >> 16) & 0xFF));

        buffer.push_back(
            static_cast<std::uint8_t>((unsignedValue >> 8) & 0xFF));

        buffer.push_back(
            static_cast<std::uint8_t>(unsignedValue & 0xFF));
    }

    void writeBool(
        std::vector<std::uint8_t>& buffer,
        bool value)
    {
        buffer.push_back(value ? 1 : 0);
    }

    void writeUint16(
        std::vector<std::uint8_t>& buffer,
        std::uint16_t value)
    {
        buffer.push_back(
            static_cast<std::uint8_t>((value >> 8) & 0xFF));

        buffer.push_back(
            static_cast<std::uint8_t>(value & 0xFF));
    }
}

std::vector<std::uint8_t> buildActivationPacket(
    const StreamOptions& options)
{
    std::vector<std::uint8_t> packet;

    packet.reserve(41);

    // Packet type: ACTIVATION
    packet.push_back(0x02);

    // Stream options
    writeInt32(packet, options.fps);

    writeInt32(packet, options.width);
    writeInt32(packet, options.height);

    writeInt32(packet, options.camera);

    writeBool(packet, options.adaptiveBitrate);

    writeInt32(packet, options.bitrate);
    writeInt32(packet, options.minBitrate);
    writeInt32(packet, options.maxBitrate);

    writeBool(packet, options.stabilization);
    writeBool(packet, options.flash);
    writeBool(packet, options.h265);

    writeInt32(packet, options.focusMode);

    // No correction/effect filters for initial activation.
    writeUint16(packet, 0);

    // Empty active effect filter.
    writeUint16(packet, 0);

    return packet;
}

File: android/app/src/main/java/com/darusc/vcamdroid/rtsp/StreamOptions.kt
package com.darusc.vcamdroid.rtsp

import com.darusc.vcamdroid.video.filters.FilterAdjuster
import com.darusc.vcamdroid.video.filters.FilterRepository
import com.darusc.vcamdroid.video.filters.custom.HorizontalFlipFilterRender
import com.darusc.vcamdroid.video.filters.custom.VerticalFlipFilterRender
import com.pedro.encoder.input.gl.render.filters.BaseFilterRender
import com.pedro.encoder.input.video.CameraHelper
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets

data class StreamOptions(
    var camera: CameraHelper.Facing = CameraHelper.Facing.BACK,
    var bitrate: Int = 4000 * 1024,
    var adaptiveBitrateMin: Int = 500 * 1024,
    var adaptiveBitrateMax: Int = 25000 * 1024,
    var adaptiveBitrateEnabled: Boolean = false,
    var width: Int = 640,
    var height: Int = 480,
    var fps: Int = 30,
    var rotation: Int = 0,
    var activeEffectFilter: BaseFilterRender? = null,
    var activeCorrectionFilters: MutableSet<BaseFilterRender> = HashSet(),
    var stabilization: Boolean = false,
    var focusMode: Int = 0,
    var h265Enabled: Boolean = false,
    var vFlipFilter: VerticalFlipFilterRender? = null,
    var hFlipFilter: HorizontalFlipFilterRender? = null
) {
    enum class FlipAxis { HORIZONTAL, VERTICAL }

    companion object {
        fun deserialize(data: ByteArray): StreamOptions {
            val buffer = ByteBuffer.wrap(data) // Defaults to Big Endian

            // Skip over the first byte which is packet type
            buffer.get()

            // 1. Fixed Fields
            val fps = buffer.getInt()
            val width = buffer.getInt()
            val height = buffer.getInt()

            val camera = if (buffer.getInt() == 1) CameraHelper.Facing.BACK else CameraHelper.Facing.FRONT
            val adaptiveBitrate = buffer.get().toInt() == 1

            val bitrate = buffer.getInt()
            val minBitrate = buffer.getInt()
            val maxBitrate = buffer.getInt()

            val stabilization = buffer.get().toInt() == 1
            val flash = buffer.get().toInt() == 1
            val h265 = buffer.get().toInt() == 1

            val focusMode = buffer.getInt()

            fun readString(): String {
                val len = buffer.short.toInt() and 0xFFFF // Unsigned short
                val bytes = ByteArray(len)
                buffer.get(bytes)
                return String(bytes, StandardCharsets.UTF_8)
            }

            // 2. Filter Sliders Map
            val filterCount = buffer.short.toInt() and 0xFFFF
            val filtersMap = mutableSetOf<BaseFilterRender>()
            repeat(filterCount) {
                val fname = readString()
                val fvalue = buffer.getInt()
                FilterRepository.create(fname)?.let {
                    FilterAdjuster.adjust(it, fvalue)
                    filtersMap.add(it)
                }
            }

            // 3. Active Filters Map
            val activeEffectFilter = readString()

            return StreamOptions(
                camera,
                bitrate,
                minBitrate,
                maxBitrate,
                adaptiveBitrate,
                width,
                height,
                fps,
                0,
                FilterRepository.create(activeEffectFilter),
                filtersMap,
                stabilization,
                focusMode,
                h265,
            )
        }
    }
}